keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,316 @@
+
+Luckily, there should only be two such cases, and in normal operation, only one would happen frequently:
+
+- The first, and most clear case of a container or image changing state without the Kubelet triggering that state change is when a container stops. Containers can exit gracefully, or be OOM killed, and the Kubelet would not know.
+   - We will also introduce events when the container is created as well as is started. This will help us reduce the relisting that takes placed while the kubelet waits for the container to start.
+   - Although kubelet initiates the container deletion, for sake of increased validation we are also introducing the event to denote that from the runtime.
+- The second, and less likely case is when another entity comes and changes the state of the node.
+	- For container related events (such as a container creating, starting, stopping or being killed), this can appear as a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) manually, or even using the runtime directly.
+	- For image related events (such as an image being pulled, or removed), this can also appear from a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md), or using various image moving tools (podman, nerdctl, skopeo, etc) to move images independent of the Kubelet.
+
+The Kubelet currently covers each of thse cases quite easily: by listing all of the resources on the node, it will have an accurate picture after the amount of time of its [poll interval](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162). For each of these cases, a new CRI-based events API can be made, using [gRPC server streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). This way, the entity closest to the activity of the containers, images and pods (the CRI implementation) can be responsible for informing the Kubelet of their behavior directly.
+
+
+### User Stories
+
+- As a cluster administrator I want to enable `Evented PLEG` feature of the kubelet for better performance.
+   - They will have to set a kubelet argument as mentioned [here](###New-Kubelet-Parameter).
+
+
+### Risks and Mitigations
+
+PLEG is very core to the container status handling in the kubelet. Hence any miscalculation there would result in unpredictible behaviour not just for the node but for an entire cluster.
+
+To reduce the risk of regression, this feature initially will be available only as an opt-in. As discussed in the section [New Kubelet Parameter](###New-Kubelet-Parameter) unless this parameter is exclusively set, the kubelet PLEG will continue to function using frequent relisting as it's doing it today. If after enabling this feature, users notice any discrepancy in container statuses they can restart the kubelet either by omitting that option or setting it to `false` explicitly.
```

Comment:
We can ensure that the existing tests function and add any addtional tests as needed to mitigate risks.